packages/dds/tree/src/shared-tree/transactionTypes.ts

```diff
@@ -0,0 +1,123 @@
+ * A requirement for a SharedTree transaction to succeed.
+ * @remarks Transaction constraints are useful for validating that the state of the tree meets some requirement when a transaction runs.
+ * In general, when running a transaction a client can validate their tree state in whatever way they wish and decide to either proceed with the transaction or not.
+ * However, they cannot know what the tree state will be when the transaction is _sequenced_.
+ * There may have been any number of edits from other clients that get sequenced before the transaction is eventually sequenced.
+ * Constraints provide a way to validate the tree state after the transaction has been sequenced and abort the transaction if the constraints are not met.
+ * All clients will validate the constraints of a transaction when it is sequenced, so all clients will agree on whether the transaction succeeds or not.
+ * @public
+ */
+export type TransactionConstraint = NodeInDocumentConstraint; // TODO: Add more constraint types here
+
+/**
+ * A transaction {@link TransactionConstraint | constraint} which requires that the given node exists in the tree.
+ * @remarks The node must be in the document (its {@link TreeStatus | status} must be {@link TreeStatus.InDocument | InDocument}) to qualify as "existing".
+ * @public
+ */
+export interface NodeInDocumentConstraint {
+	readonly type: "nodeInDocument";
+	readonly node: TreeNode;
+}
+
+/**
+ * The status of a the transaction callback in the {@link RunTransaction | RunTransaction} API.
+ * @alpha
+ */
+export type TransactionCallbackStatusExt<TSuccessValue, TFailureValue> = (
+	| {
+			/** Indicates that the transaction callback ran successfully. */
+			rollback?: false;
+			/** The user defined value when the transaction ran successfully. */
+			value: TSuccessValue;
+	  }
+	| {
+			/** Indicates that the transaction callback failed and the transaction should be rolled back. */
+			rollback: true;
+			/** The user defined value when the transaction failed. */
+			value: TFailureValue;
+	  }
+) & {
+	// Todo: For demonstration only. To be removed.
+	undoPreconditions?: readonly TransactionConstraint[];
+};
+
+/**
+ * The status of a the transaction callback in the {@link RunTransaction | RunTransaction} API. This is the
+ * same as {@link TransactionCallbackStatusExt} but with the `value` field omitted. This is useful when
+ * the user doesn't want to return any value from the transaction callback.
+ * @alpha
+ */
+export type TransactionCallbackStatus = Omit<
+	TransactionCallbackStatusExt<unknown, unknown>,
+	"value"
+>;
+
+/**
+ * The result of the {@link RunTransaction | RunTransaction} API when it was successful.
+ * @alpha
+ */
+export interface TransactionResultSuccess<TSuccessValue> {
+	/** Indicates that the transaction was successful. */
+	success: true;
+	/** The user defined value when the transaction was successful. */
+	value: TSuccessValue;
+	// Todo: For demonstration only. To be removed.
+	commitId?: string;
```

Comment:
Just putting a comment here so we don't forget to remove :)

Reply:
removed it.